transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,266 @@
+        return bs.connect();
+    }
+
+    @Override
+    public final Future<Void> release(Channel channel) {
+        return release(channel, channel.eventLoop().<Void>newPromise());
+    }
+
+    @Override
+    public Future<Void> release(final Channel channel, final Promise<Void> promise) {
+        checkNotNull(channel, "channel");
+        checkNotNull(promise, "promise");
+        try {
+            EventLoop loop = channel.eventLoop();
+            if (loop.inEventLoop()) {
+                doReleaseChannel(channel, promise);
+            } else {
+                loop.execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        doReleaseChannel(channel, promise);
+                    }
+                });
+            }
+        } catch (Throwable cause) {
+            closeAndFlail(channel, cause, promise);
+        }
+        return promise;
+    }
+
+    private void doReleaseChannel(Channel channel, Promise<Void> promise) {
+        assert channel.eventLoop().inEventLoop();
+        // Remove the POOL_KEY attribute from the Channel and check if it was acquired from this pool, if not fail.
+        if (channel.attr(POOL_KEY).getAndSet(null) != this) {
+            closeAndFlail(channel,
+                          // Better include a stracktrace here as this is an user error.
+                          new IllegalStateException("Channel " + channel + " was not acquired from this ChannelPool"),
+                          promise);
+        } else {
+            try {
+                if (offerChannel(channel)) {
+                    handler.channelReleased(channel);
+                    promise.setSuccess(null);
+                } else {
+                    closeAndFlail(channel, FULL_EXCEPTION, promise);
+                }
+            } catch (Throwable cause) {
+                closeAndFlail(channel, cause, promise);
+            }
+        }
+    }
+
+    private static void closeChannel(Channel channel) {
+        channel.attr(POOL_KEY).getAndSet(null);
+        channel.close();
+    }
+
+    private static void closeAndFlail(Channel channel, Throwable cause, Promise<?> promise) {
+        closeChannel(channel);
+        promise.setFailure(cause);
+    }
+
+    /**
+     * Poll a {@link Channel} out of the internal storage to reuse it. This will return {@code null} if no
+     * {@link Channel} is ready to be reused.
+     *
+     * Sub-classes may override {@link #pollChannel()} and {@link #offerChannel(Channel)}. Be aware that
+     * implementations of these methods needs to be thread-safe!
+     */
+    protected Channel pollChannel() {
+        return deque.pollLast();
+    }
+
+    /**
+     * Offer a {@link Channel} back to the internal storage. This will return {@code true} if the {@link Channel}
+     * could be added, {@code false} otherwise.
+     *
+     * Sub-classes may override {@link #pollChannel()} and {@link #offerChannel(Channel)}. Be aware that
```

Comment:
yeah both need to be thread-safe as these are called from different EventLoops (the EventLoop of the Channels). We could just do it with the same each time but I'm not sure yet if this is a good idea.